itou/eligibility/models/iae.py

```diff
@@ -75,25 +75,28 @@ def last_considered_valid(self, job_seeker, for_siae=None):
         diagnosis already exists.
         """
 
-        last = None
         query = self.for_job_seeker(job_seeker).order_by("created_at")
+        if not job_seeker.has_valid_common_approval:
+            query = query.valid()
+        # Otherwise, a diagnosis is considered valid for the duration of an
+        # approval, we just retrieve the last one no matter if it's valid or
+        # not.
 
-        # A diagnosis is considered valid for the duration of an approval,
-        # we just retrieve the last one no matter if it's valid or not.
-        if job_seeker.has_valid_common_approval:
-            last = query.by_author_kind_prescriber().last()
-            if not last and for_siae:
-                last = query.authored_by_siae(for_siae).last()
-            if not last:
-                # Deals with cases from the past (when there was no restriction).
-                last = query.last()
+        filter = models.Q(author_kind=AuthorKind.PRESCRIBER)
+        if for_siae:
+            filter |= models.Q(author_siae=for_siae)
 
-        # Otherwise, search only in "non expired" diagnosis.
+        last = None
+        for diagnosis in query.filter(filter):
+            # Prescriber diagnosis have precedence over SIAE diagnosis.
+            if diagnosis.author_kind == AuthorKind.PRESCRIBER:
+                last = diagnosis
+            elif for_siae and not last:
+                last = diagnosis
         else:
-            last = query.valid().by_author_kind_prescriber().last()
-            if not last and for_siae:
-                last = query.valid().authored_by_siae(for_siae).last()
-
+            # Deals with cases from the past (when there was no restriction).
+            if job_seeker.has_valid_common_approval:
+                last = query.last()
```

Comment:
Je trouve ce code complexe à lire (je ne comprends pas à quoi sert le `for/else` vu qu'il n'y a pas de `break` dans la boucle) et j'ai l'impression que dans le cas où il y a un pass en cours, on va faire la requête `query.filter(...)` et `query.last()` donc il va y avoir 2 requêtes.
Il faudrait ajouter ce cas dans le test.

Reply:
Malheureusement, les trois cas correspondent bien à un diagnostic pour un PASS en cours de validité (et jusqu’à 2026).

Reply:
Je vais peut-être bien réussir à dégager ce cas finalement : https://itou-inclusion.slack.com/archives/C01AQKD7MAN/p1690820807059479